src/sentry/event_manager.py

```diff
@@ -295,15 +320,90 @@ class InvalidTimestamp(Exception):
+    """
 
     def __init__(self, data, version='5'):
-        self.data = CanonicalKeyDict(data)
+        if not isinstance(data, (bytes, six.text_type)):
+            data = CanonicalKeyDict(data)
+        self._data = data
         self.version = version
 
-    def normalize(self, request_env=None):
-        request_env = request_env or {}
-        data = self.data
+    def decode(self, content_encoding, helper):
+        data = self._data
+        if isinstance(data, six.binary_type):
+            if content_encoding == 'gzip':
+                data = helper.decompress_gzip(data)
+            elif content_encoding == 'deflate':
+                data = helper.decompress_deflate(data)
+            elif data[0] != b'{':
+                data = helper.decode_and_decompress_data(data)
+            else:
+                data = helper.decode_data(data)
+        if isinstance(data, six.text_type):
+            data = helper.safely_load_json_string(data)
+
+        self._data = CanonicalKeyDict(data)
+
+    def process_csp_report(self, client_ip=None, user_agent=None):
+        """Only called from the CSP report endpoint."""
+        data = self._data
+
+        try:
+            interface = get_interface(data.pop('interface'))
+            report = data.pop('report')
+        except KeyError:
+            raise APIForbidden('No report or interface data')
+
+        # To support testing, we can either accept a buillt interface instance, or the raw data in
+        # which case we build the instance ourselves
+        try:
+            instance = report if isinstance(report, interface) else interface.from_raw(report)
+        except jsonschema.ValidationError as e:
+            raise APIError('Invalid security report: %s' % str(e).splitlines()[0])
+
+        def clean(d):
+            return dict(filter(lambda x: x[1], d.items()))
+
+        data.update({
+            'logger': 'csp',
+            'message': instance.get_message(),
+            'culprit': instance.get_culprit(),
+            instance.get_path(): instance.to_json(),
+            'tags': instance.get_tags(),
+            'errors': [],
+
+            'sentry.interfaces.User': {
+                'ip_address': client_ip,
+            },
+
+            # Construct a faux Http interface based on the little information we have
+            # This is a bit weird, since we don't have nearly enough
+            # information to create an Http interface, but
+            # this automatically will pick up tags for the User-Agent
+            # which is actually important here for CSP
+            'sentry.interfaces.Http': {
+                'url': instance.get_origin(),
+                'headers': clean({
+                    'User-Agent': user_agent,
+                    'Referer': instance.get_referrer(),
+                })
+            },
+        })
+
+        self._data = data
+
+    def normalize(self, project=None, key=None, auth=None, client_ip=None):
```

Comment:
These kwargs need to be roughly in sync with what @tonyo's normalization script is passing to rust later. Since we need to be able to enumerate these, I'd be in favor of typing this out in a class and mirroring it in libsemaphore. FWIW, the same can then be passed to `process_csp_report`.

Reply:
Those kwargs likely need to be part of `__init__` anyway

Reply:
I moved most kwargs to `__init__`, I think that solves the problem